tls/extensions/s2n_client_supported_versions.c

```diff
@@ -71,6 +71,13 @@ int s2n_extensions_client_supported_versions_process(struct s2n_connection *conn
         uint8_t client_version_parts[S2N_TLS_PROTOCOL_VERSION_LEN];
         GUARD(s2n_stuffer_read_bytes(extension, client_version_parts, S2N_TLS_PROTOCOL_VERSION_LEN));
 
+        /* If the client version is outside of our supported versions, then ignore the value.
+         * S2N does not support SSLv2 except for upgrading connections. Since this extension is
+         * a TLS1.3 extension, we will skip any SSLv2 values. */
+        if (client_version_parts[0] != 3 || client_version_parts[1] > 4) {
```

Comment:
instead of adding more checks, i would simplify version conversions to something like

```c
uint16_t s2n_tls_version_from_wire_value(uint16_t wire) {
    switch (wire) {
    case 0x0200: return S2N_SSLv2;
    case 0x0300: return S2N_SSLv3;
    case 0x0301: return S2N_TLS10;
    case 0x0302: return S2N_TLS11;
    case 0x0303: return S2N_TLS12;
    case 0x0304: return S2N_TLS13;
    default: return S2N_UNKNOWN_PROTOCOL_VERSION;
    }
}

int s2n_tls_version_to_wire_value(uint16_t version) {
    switch (version) {
    case S2N_SSLv2: return 0x0200;
    case S2N_SSLv3: return 0x0300;
    case S2N_TLS10: return 0x0301;
    case S2N_TLS11: return 0x0302;
    case S2N_TLS12: return 0x0303;
    case S2N_TLS13: return 0x0304;
    }

    return S2N_FAILURE;
}
```

Reply:
In this case the switch statements compile down to several checks, which is what I wanted to avoid. We could order the cases to start with 0x0304 though, which will handle the most common case in this extension.

I agree that having the conversion would be nice though, but I would want to use it everywhere. Perhaps this is a new issue?

Reply:
https://github.com/awslabs/s2n/issues/1634 created for this update.

Reply:
While this switch statement may not be faster than the math statement or switching over an enum, I wouldn't worry too much about the ordering, the compiler optimizes that so a constant amount comparisons is made before a jump. I think it all ends up being equivalent or better, since it reduces additional checks and branches for edge cases.